File: src/store/book-store.ts
import {
  BookStatus,
  BookNotFoundError,
  DuplicateBookIdError,
  createBook,
  type Book,
} from '../domain/book.js';

function clone(book: Book): Book {
  return { ...book };
}

/**
 * 进程内图书集合，按 `id` 索引。
 * 只提供数据读写，不做借阅/归还的业务校验；见 docs/designs/FP-001。
 */
export class BookStore {
  private readonly books = new Map<string, Book>();

  /** 新增图书；编号重复或字段非法时抛错。 */
  add(book: Book): Book {
    const normalized = createBook(book.id, book.title, {
      status: book.status,
      borrower: book.borrower,
    });

    if (this.books.has(normalized.id)) {
      throw new DuplicateBookIdError(normalized.id);
    }

    this.books.set(normalized.id, normalized);
    return clone(normalized);
  }

  /** 便捷新增：以编号 + 名称建一本默认可借的图书。 */
  create(id: string, title: string): Book {
    return this.add(createBook(id, title));
  }

  /** 返回全部图书，按插入顺序。 */
  listAll(): Book[] {
    return [...this.books.values()].map(clone);
  }

  /** 按编号查找；不存在返回 `undefined`。 */
  find(id: string): Book | undefined {
    const book = this.books.get(id);
    return book ? clone(book) : undefined;
  }

  /** 更新状态与借阅人；编号不存在时抛 `BookNotFoundError`。 */
  setStatus(
    id: string,
    status: BookStatus,
    borrower: string | null = null,
  ): Book {
    const book = this.books.get(id);
    if (!book) {
      throw new BookNotFoundError(id);
    }

    book.status = status;
    book.borrower = status === BookStatus.Available ? null : (borrower ?? null);
    return clone(book);
  }

  /** 清空集合（重置/测试用）。 */
  clear(): void {
    this.books.clear();
  }
}

/** 进程内单例集合，供各功能任务共享同一份内存数据。 */
export const bookStore = new BookStore();

File: library/book_store.py
"""Minimal in-memory book collection — the task-card §6 fallback for FP-001.

FP-001 (§3.1) is implemented in TypeScript (``src/store/book-store.ts``);
a Python process cannot read it. Per task FP-010 §6, this module is the
"内联最小图书集合" stand-in so the Python CLI (FP-005) can run the list
command standalone. It intentionally implements only the surface needed
by FP-010 plus enough to simulate borrow/return in tests — no validation,
error types or uniqueness rules. Replace it with the real FP-001 port
when one exists.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional

AVAILABLE = "available"
BORROWED = "borrowed"


@dataclass
class Book:
    """A book row as consumed by the list command (FP-001 §3.1 shape)."""

    id: str
    title: str
    status: str = AVAILABLE
    borrower: Optional[str] = None


class BookStore:
    """Insertion-ordered in-memory collection of :class:`Book`."""

    def __init__(self) -> None:
        self._books: Dict[str, Book] = {}

    def add(self, book: Book) -> Book:
        if book.id in self._books:
            raise ValueError(f"Duplicate book id: {book.id}")
        self._books[book.id] = book
        return book

    def create(self, book_id: str, title: str) -> Book:
        return self.add(Book(id=book_id, title=title))

    def list_all(self) -> List[Book]:
        return list(self._books.values())

    def find(self, book_id: str) -> Optional[Book]:
        return self._books.get(book_id)

    def set_status(
        self, book_id: str, status: str, borrower: Optional[str] = None
    ) -> Book:
        book = self._books.get(book_id)
        if book is None:
            raise KeyError(book_id)
        book.status = status
        book.borrower = None if status == AVAILABLE else borrower
        return book

    def clear(self) -> None:
        self._books.clear()


#: Process-wide collection shared by the CLI commands.
book_store = BookStore()

#: Sample seed: B001《三体》/ B002《活着》/ B003《百年孤独》, all available.
SAMPLE_BOOKS = (
    ("B001", "三体"),
    ("B002", "活着"),
    ("B003", "百年孤独"),
)


def seed_sample_books(store: Optional[BookStore] = None) -> List[Book]:
    """Load the sample books into ``store`` (default: the shared singleton).

    Idempotent: ids already present are skipped, so a running status is
    never overwritten. Returns the books actually added this call.
    """
    target = book_store if store is None else store
    added: List[Book] = []
    for book_id, title in SAMPLE_BOOKS:
        if target.find(book_id) is None:
            added.append(target.create(book_id, title))
    return added
